src/ripple/basics/safe_cast.h

```diff
@@ -0,0 +1,78 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2018 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_BASICS_SAFE_CAST_H_INCLUDED
+#define RIPPLE_BASICS_SAFE_CAST_H_INCLUDED
+
+#include <type_traits>
+
+namespace ripple {
+
+// safe_cast adds compile-time checks to a static_cast to ensure that
+// the destination can hold all values of the source.  This is particularly
+// handy when the source or destination is an enumeration type.
+
+template <class Dest, class Src>
+inline
+constexpr
+std::enable_if_t
+<
+    std::is_integral<Dest>::value && std::is_integral<Src>::value,
+    Dest
+>
+safe_cast(Src s) noexcept
+{
+    static_assert(std::is_signed<Dest>::value || std::is_unsigned<Src>::value,
+        "Can not cast signed to unsigned");
+    constexpr unsigned not_same = std::is_signed<Dest>::value !=
+                                  std::is_signed<Src>::value;
+    static_assert(sizeof(Dest) >= sizeof(Src) + not_same,
+        "Destination is too small to hold all values of source");
+    return static_cast<Dest>(s);
+}
```

Comment:
Thanks for the thorough testing!

Reply:
Controversial opinion: I think we should include @scottschurr's tests. Reasons:
1. It verifies that everything is working as expected on all of our current supported platforms.
2. All the types listed have explicit sizes, so if a test did fail, there would be something unusual enough that we would _really_ want to know about it.
3. I'd be willing to bet that we have code all over the place that assumes these types have the size relationships that they do, even without explicit casting. If that were ever to not be true, similar to 2, it would be unusual enough that we'd want to know about it.

Reply:
The only way I know of to include these tests in our unit tests is to move the `static_assert`s to the `enable_if`s.  This will sacrifice the `static_assert` messages which will fire as we develop, even though `safe_cast` is developed and proven only once.

Though if you know of a way to apply these tests and not sacrifice the `static_assert` messages, that would be the best of both worlds.

Reply:
As an example of the `static_assert` messages adding usability:

Consider the scenario where we change the underlying type of `Type` to `std::int8_t` because we want to add a negative member:

https://github.com/ripple/rippled/blob/develop/src/ripple/conditions/Condition.h#L37

This will cause an error far away, here:

https://github.com/ripple/rippled/pull/2814/commits/7d0c01785fa595ad640fecd810529cd345ebd335#diff-9634052245df9b2eee790a9804088d25R123

If we add the tests, the error message will be something like:

`No matching call to safe_cast.`

As it is now, the error message is:

`Cannot cast signed to unsigned`

Reply:
I agree with Howard that the `static_assert`s significantly improve usability.

Something I've wished for but have never seen is a CI phase that compiles a bunch of files with the expectation that each compilation will fail.  Then all of the cases that we expect to fail to compile could be put in separate files and run during that CI phase.  I've never seen any  framework that supports that model.  So I shrug and do occasional manual testing.

Reply:
I wrote myself such a test framework for libcxx and date.  But it doesn't fit well into the rippled model where tests are part of the application binary.  What I wrote is a simplistic bash shell combined with a policy that tests which are expected to fail have filenames that end with `.fail.cpp`.  The bash script just runs over them and marks them as fail if they compile.  It doesn't even have the ability to check that they failed for the right reason.